src/index.tsx

```diff
@@ -52,211 +52,235 @@ interface OTPInputProps {
   skipDefaultStyles?: boolean; // TODO: Remove in next major release
 }
 
+interface OTPRef {
+  focus: () => void;
+}
+
 const isStyleObject = (obj: unknown) => typeof obj === 'object' && obj !== null;
 
-const OTPInput = ({
-  value = '',
-  numInputs = 4,
-  onChange,
-  onPaste,
-  renderInput,
-  shouldAutoFocus = false,
-  inputType = 'text',
-  renderSeparator,
-  placeholder,
-  containerStyle,
-  inputStyle,
-  skipDefaultStyles = false,
-}: OTPInputProps) => {
-  const [activeInput, setActiveInput] = React.useState(0);
-  const inputRefs = React.useRef<Array<HTMLInputElement | null>>([]);
-
-  const getOTPValue = () => (value ? value.toString().split('') : []);
-
-  const isInputNum = inputType === 'number' || inputType === 'tel';
-
-  React.useEffect(() => {
-    inputRefs.current = inputRefs.current.slice(0, numInputs);
-  }, [numInputs]);
-
-  React.useEffect(() => {
-    if (shouldAutoFocus) {
-      inputRefs.current[0]?.focus();
-    }
-  }, [shouldAutoFocus]);
-
-  const getPlaceholderValue = () => {
-    if (typeof placeholder === 'string') {
-      if (placeholder.length === numInputs) {
-        return placeholder;
+const OTPInput = React.forwardRef<OTPRef, OTPInputProps>(
+  (
+    {
+      value = '',
+      numInputs = 4,
+      onChange,
+      onPaste,
+      renderInput,
+      shouldAutoFocus = false,
+      inputType = 'text',
+      renderSeparator,
+      placeholder,
+      containerStyle,
+      inputStyle,
+      skipDefaultStyles = false,
+    },
+    ref
+  ) => {
+    const [activeInput, setActiveInput] = React.useState(0);
+    const inputRefs = React.useRef<Array<HTMLInputElement | null>>([]);
+
+    const getOTPValue = () => (value ? value.toString().split('') : []);
+
+    const isInputNum = inputType === 'number' || inputType === 'tel';
+
+    React.useEffect(() => {
+      inputRefs.current = inputRefs.current.slice(0, numInputs);
+    }, [numInputs]);
+
+    React.useEffect(() => {
+      if (shouldAutoFocus) {
+        inputRefs.current[0]?.focus();
       }
+    }, [shouldAutoFocus]);
+
+    React.useImperativeHandle(
+      ref,
+      () => ({
+        focus() {
+          const inputRef = inputRefs.current[value.length] ?? inputRefs.current[0];
```

Comment:
Also why are we relying on the length of the value, any reason for not doing it like this:
```suggestion
          const inputRef = inputRefs.current[activeInput]
```

Reply:
I tried using activeInput but the behaviour was not what seems like the best UX. Sometimes, the active input is not the first unfilled input, but a filled input before. So you would overwrite the last value of the otp when clicking focus. Using the value length allows to always focus the first unfilled input.